File: python/i20.subroutine.py
#!/usr/bin/python

pond = 500
captured = 0

def fish():
    global captured # explicitly declare we'll use global caputured
    pond = 500      # must declare some number before subtracting from it
    pond -= 150     # now subtract from local pond
    captured += 150 # add to the fish captured
    
    
print "We have {} in this pond.".format(pond)    

fish()  # get some fish
print "Fishing from a local pond... We now have {} in the main pond.".format(pond)

fish()  # get some fish
print "Fishing from a local pond... We now have {} in the main pond.".format(pond)

fish()  # get some fish
print "Fishing from a local pond... We now have {} in the main pond.".format(pond)

print "We now have a total of {} fish".format(captured)

File: compiled_lang/java/src/i20.subroutine.java
class I20Subroutine {
    static int pond = 500; // never mutated - fish() only touches its own local copy
    static int captured = 0;

    static void fish() {
        int pond = 500; // shadows the static field for the rest of this method
        pond -= 150;
        captured += 150;
    }

    public static void main(String[] args) {
        System.out.println("We have " + pond + " in this pond.");

        fish();
        System.out.println("Fishing from a local pond... We now have " + pond + " in the main pond.");

        fish();
        System.out.println("Fishing from a local pond... We now have " + pond + " in the main pond.");

        fish();
        System.out.println("Fishing from a local pond... We now have " + pond + " in the main pond.");

        System.out.println("We now have a total of " + captured + " fish captured");
    }
}
